### PR TITLE
[bluetooth.generic] Added support for generic bluetooth devices

### PR DESCRIPTION
**Pull request by cpmeister**
This is the long awaited update to the bluetooth binding that adds support for generic bluetooth devices.
Generic bluetooth devices will dynamically generate their channels based upon the available gatt characteristics of that device.

I built upon the work of @vkolotov from his 3rd party binding by using some of his libraries and a couple of classes from his binding codebase. I've had to modify his code to get it to work but I've tried to give credit where credit is due. But if I missed something or didn't properly add license info please let me know.

Before this PR is merged, I'd like additional units to be supported in the core. 
https://github.com/openhab/openhab-core/pull/1672

But that is more of a desire than a requirement, this can be merged without that PR and I can complete
unit support for this binding after it is merged.

Signed-off-by: Connor Petty <mistercpp2000+gitsignoff@gmail.com>

### CODE REVIEW

**Comment by cpmeister** [comment]
> Well done @cpmeister, I have no issues regarding using bits and pieces from the 3rd party binding and the GATT parser lib. If you need any changes in those libs, feel free to create a PR.

I greatly appreciate the work you put into the gatt-parser library! It really laid the groundwork for handling channel<->characteristic communication and I don't think I would have the patience to create that myself.

I might create a PR to add the osgi headers to your manifest file but it might not be needed.

**Comment by cpmeister** [comment]
I went ahead and put the generic bluetooth device support as a separate bluetooth extension, should make it easier to deal with dependency handling as well as minimizing dependency bloat for the other extensions.

Even without all the units being added in the core, I have workaround code that should more than suffice for the moment.
I feel this is ready for review.

File: bundles/org.openhab.binding.bluetooth.generic/pom.xml
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.generic/pom.xml:35]: Are these used anywhere?

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.bluetooth.generic/README.md
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/README.md]: done

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java:83]: This is a placeholder until the additional units in openhab core get merged. Until that gets merged I won't add support for quantity types in this binding. I'd prefer to leave this TODO here as it is something that can't be immediately addressed but also something that will be done.

File: bundles/org.openhab.binding.bluetooth.generic/pom.xml
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/pom.xml:35]: They aren't used in this binding, but they are transitive dependencies for the gatt-parser library and in order for them to be packaged into this binding's jar I have to explicitly label them as compile dependencies in the pom.

Comment (fwolter) [reply bundles/org.openhab.binding.bluetooth.generic/pom.xml:35]: Are you sure? In the GATT POM, only commons-beanutils is listed: https://github.com/sputnikdev/bluetooth-gatt-parser/blob/master/pom.xml

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericDiscoveryParticipant.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericDiscoveryParticipant.java:105]: Generic bluetooth devices were always intended to be the fallback. Originally this binding was integrated into the bluetooth binding code and used as the fallback during discovery, but I separated it into its own binding so I had to add this `order` method to the discovery participant class so that the generic bluetooth devices could still act as the fallback. If there is ever a time that something else should be the fallback (I can't imagine what) then we can change the  `order` value of generic devices.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:358]: Even I don't know what to do about this since the TODO is from @vkolotov's code that I copied over. It is unknown at this time if/how to implement this.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:249]: I'm not knowledgable enough about bluetooth atm to know how to implement this. This TODO is from the original copied code. I don't know if there will be enough demand to implement this but I find that the TODO is a nice placeholder for knowing where code would need to be added in order to implement the feature.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:125]: whoops, it should be. I accidentally left that at a higher level during debugging.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java:67]: This whole class is incomplete due to the lack of proper support of the units in the core. In fact a lot of the self made units are meant to be removed from this class once those units are supported in the core. I'll do what I can to cleanup what is already here though.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java:206]: It isn't but it is an important identifier nonetheless. I added this field to be consistent with the other bluetooth protocol classes included in the bluetooth binding.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java:204]: done

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java:197]: done

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java]: removed

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java]: done

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java]: So would I, there are a couple of PRs for it (https://github.com/openhab/openhab-core/pull/1672 and https://github.com/openhab/openhab-core/pull/1664) that are currently languishing in PR purgatory.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java]: It was used to generate most of this file. But I don't need it anymore so I'll remove it.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnit.java]: done

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java]: Thanks for spotting this.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:243]: done

File: bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/discovery/BluetoothDiscoveryParticipant.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/discovery/BluetoothDiscoveryParticipant.java:98]: fixed

File: bundles/org.openhab.binding.bluetooth.generic/pom.xml
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/pom.xml:35]: I am sure, because without including it here karaf would throw a fit during compilation wanting me to add commons-collections to the feature.xml. I didn't want to do that so I added it as a compile dependency instead.

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/pom.xml:35]: I looked further into it and found that the commons-collections library is a dependency for the commons-beanutils library.

**Review by fwolter — APPROVED** [review]
LGTM. You need to run `mvn spotless:apply`.

**Comment by cpmeister** [comment]
@openhab/add-ons-maintainers Can I please get a second review for this, I'd really like this to be included in the 3.0 release since I keep getting inquiries as to when it will be merged.

File: bundles/org.openhab.binding.bluetooth.generic/README.md
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/README.md:3]:
```suggestion
This binding adds support for devices that expose [Bluetooth Generic Attributes (GATT)](https://www.bluetooth.com/specifications/gatt/)
```

File: bundles/org.openhab.binding.bluetooth.generic/README.md
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/README.md:7]:
```suggestion
Only a single thing type is added by this binding:
```

File: bundles/org.openhab.binding.bluetooth.generic/README.md
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/README.md:15]:
```suggestion
As any other Bluetooth device, generic bluetooth devices are discovered automatically by the corresponding bridge.
```

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java:44]: Are there only minor modifications or larger ones?

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java:99]:
```suggestion
                state = OnOffType.from(holder.getBoolean());
```

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:71]:
```suggestion
    private @Nullable ScheduledFuture<?> readCharacteristicJob = null;
```

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:111]:
```suggestion
        ScheduledFuture<?> future = readCharacteristicJob;
```

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java:357]: I would suggest to remove it then.

File: bundles/org.openhab.binding.bluetooth.generic/src/test/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnitTest.java
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.generic/src/test/java/org/openhab/binding/bluetooth/generic/internal/BluetoothUnitTest.java:18]: Please fix author

**Review by J-N-K — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/BluetoothChannelUtils.java:44]: I slightly modified how a couple of methods worked and added/removed some some other methods.

File: bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java]: I just implemented it instead.

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
Jenkins fails unrelated

**Pull request merged**
